src/algebra/star/chsh.lean

```diff
@@ -0,0 +1,265 @@
+(A₁B₀_commutes : A₁ * B₀ = B₀ * A₁)
+(A₁B₁_commutes : A₁ * B₁ = B₁ * A₁)
+
+variables {R : Type u}
+
+/--
+Given a CHSH tuple (A₀, A₁, B₀, B₁) in a *commutative* ordered *-algebra over ℝ,
+`A₀ * B₀ + A₀ * B₁ + A₁ * B₀ - A₁ * B₁ ≤ 2`.
+
+(We could work over ℤ[⅟2] if we wanted to!)
+-/
+lemma CHSH_inequality_of_comm
+  [ordered_comm_ring R] [star_ordered_ring R] [algebra ℝ R] [ordered_semimodule ℝ R]
+  (A₀ A₁ B₀ B₁ : R) (T : is_CHSH_tuple A₀ A₁ B₀ B₁) :
+  A₀ * B₀ + A₀ * B₁ + A₁ * B₀ - A₁ * B₁ ≤ 2 :=
+begin
+  let P := (2 - A₀ * B₀ - A₀ * B₁ - A₁ * B₀ + A₁ * B₁),
+  have i₁ : 0 ≤ P,
+  { have idem : P * P = 4 * P,
+    { -- If we had a Gröbner basis algorithm, this would be trivial.
+      -- Without one, it is somewhat tedious!
+      dsimp [P],
+      simp only [add_mul, mul_add, sub_mul, mul_sub, mul_comm, mul_assoc, add_assoc],
+      repeat { conv in (B₀ * (A₀ * B₀))
+      { rw [T.A₀B₀_commutes, ←mul_assoc B₀ B₀ A₀, ←pow_two, T.B₀_inv, one_mul], } },
+      repeat { conv in (B₀ * (A₁ * B₀))
+      { rw [T.A₁B₀_commutes, ←mul_assoc B₀ B₀ A₁, ←pow_two, T.B₀_inv, one_mul], } },
+      repeat { conv in (B₁ * (A₀ * B₁))
+      { rw [T.A₀B₁_commutes, ←mul_assoc B₁ B₁ A₀, ←pow_two, T.B₁_inv, one_mul], } },
+      repeat { conv in (B₁ * (A₁ * B₁))
+      { rw [T.A₁B₁_commutes, ←mul_assoc B₁ B₁ A₁, ←pow_two, T.B₁_inv, one_mul], } },
+      conv in (A₀ * (B₀ * (A₀ * B₁)))
+      { rw [←mul_assoc, T.A₀B₀_commutes, mul_assoc, ←mul_assoc A₀, ←pow_two, T.A₀_inv, one_mul], },
+      conv in (A₀ * (B₁ * (A₀ * B₀)))
+      { rw [←mul_assoc, T.A₀B₁_commutes, mul_assoc, ←mul_assoc A₀, ←pow_two, T.A₀_inv, one_mul], },
+      conv in (A₁ * (B₀ * (A₁ * B₁)))
+      { rw [←mul_assoc, T.A₁B₀_commutes, mul_assoc, ←mul_assoc A₁, ←pow_two, T.A₁_inv, one_mul], },
+      conv in (A₁ * (B₁ * (A₁ * B₀)))
+      { rw [←mul_assoc, T.A₁B₁_commutes, mul_assoc, ←mul_assoc A₁, ←pow_two, T.A₁_inv, one_mul], },
+      simp only [←pow_two, T.A₀_inv, T.A₁_inv],
+      simp only [mul_comm A₁ A₀, mul_comm B₁ B₀, mul_left_comm A₁ A₀, mul_left_comm B₁ B₀,
+        mul_left_comm B₀ A₀, mul_left_comm B₀ A₁, mul_left_comm B₁ A₀, mul_left_comm B₁ A₁],
+      norm_num,
+      simp only [mul_comm _ (2 : R), mul_comm _ (4 : R),
+        mul_left_comm _ (2 : R), mul_left_comm _ (4 : R)],
+      abel,
+      simp only [neg_mul_eq_neg_mul_symm, mul_one, int.cast_bit0, one_mul, int.cast_one,
+        gsmul_eq_mul, int.cast_neg],
+      simp only [←mul_assoc, ←add_assoc],
+      norm_num, },
+    have idem' : P = (1 / 4 : ℝ) • (P * P),
+    { have h : 4 * P = (4 : ℝ) • P := by simp [algebra.smul_def],
+      rw [idem, h, ←mul_smul],
+      norm_num, },
+    have sa : star P = P,
+    { dsimp [P],
+      simp only [star_add, star_sub, star_mul, star_bit0, star_one,
+        T.A₀_sa, T.A₁_sa, T.B₀_sa, T.B₁_sa, mul_comm B₀, mul_comm B₁], },
+    rw idem',
+    conv_rhs { congr, skip, congr, rw ←sa, },
+    convert smul_le_smul_of_nonneg (star_mul_self_nonneg : 0 ≤ star P * P) _,
+    { simp, },
```

Comment:
```suggestion
    { simp, },
    { apply_instance, },
```

Reply:
(I realise this will probably be refactored substantially before merge so this is hardly worth it but I guess no harm having the build passing.)

Reply:
Thanks!